conf/config-default.yaml

```diff
@@ -14,122 +14,121 @@
-                                  #  !!! So do not change it after saving your ssl, it can't decrypt the ssl keys have be saved if you change !!
-                                  #  Only use the first key to encrypt, and decrypt in the order of the array.
-
-    #fallback_sni: "my.default.domain"      # If set this, when the client doesn't send SNI during handshake, the fallback SNI will be used instead
-  enable_control: true
-  #control:
+    ssl_session_tickets: false  # If true, session tickets are used for SSL/TLS connections.
+                                # Disabled by default because it renders Perfect Forward Secrecy (FPS) 
+                                # useless. See https://github.com/mozilla/server-side-tls/issues/135.
+
+    key_encrypt_salt:           # Salt for SSL/TLS private key used with AES-128-CBC.
+      - edd1c9f0985e76a2        # Each salt value should be a hexadecimal string of length 16.
+                                # Support multiple salt values for rotation.
+                                # CAUTION: DO NOT MODIFY SALT VALUE AFTER KEY IS WRITTEN TO ETCD. 
+                                # VALUES CANNOT BE DECRYPTED OTHERWISE.
+
+    # fallback_sni: "my.default.domain"      # Fallback SNI to use if client does not send SNI during 
+                                             # the handshake.
+    
+  enable_control: true  # Control API
+  # control:
   #  ip: 127.0.0.1
   #  port: 9090
-  disable_sync_configuration_during_start: false  # safe exit. Remove this once the feature is stable
-  data_encryption:                # add `encrypt_fields = { $field },` in plugin schema to enable encryption
-    enable: false                 # if not set, the default value is `false`.
+
+  disable_sync_configuration_during_start: false  # Safe exit. TO BE REMOVED.
+
+  data_encryption:                # Encrypt fields specified in `encrypt_fields` in plugin schema.
+    enable: false
     keyring:
-      - qeddd145sfvddff3          # If not set, will save origin value into etcd.
-                                  # If set this, the keyring should be an array whose elements are string, and the size is also 16, and it will encrypt fields with AES-128-CBC
-                                  # !!! So do not change it after encryption, it can't decrypt the fields have be saved if you change !!
-                                  # Only use the first key to encrypt, and decrypt in the order of the array.
+      - qeddd145sfvddff3          # Salt for other fields encrypted with AES-128-CBC.
+                                  # Each salt value should be a hexadecimal string of length 16.
+                                  # Support multiple salt values for rotation.
+                                  # CAUTION: DO NOT MODIFY SALT VALUE AFTER CONFIGURATIONS ARE WRITTEN INTO ETCD. VALUES CANNOT BE DECRYPTED OTHERWISE.
 
```

Comment:
1. If not set, will save origin value into etcd.
2. Only use the first key to encrypt, and decrypt in the order of the array.
Why do you remove the two sentence? They are used to explain the function

Reply:
1. if the salt is nil, the encryption won't take place either? AES-128-CBC won't work without a salt?
2. They are but not very clear to me what these mean. I looked up how AES-128-CBC and key rotation work to have a high level understanding. I would either leave it for the users to learn more about AES-128-CBC and key rotation on their own or expand the explanation here, which I need help for. Feel free to provide your suggestions.

Reply:
1. yes, the encryption won't happen if without a salt
2. For the convenience of key replacement, we support using multiple keys simultaneously, encrypting only using the first key, and using the keys in sequence to decrypt (if decryption fails).

Reply:
1. Thanks. Will update accordingly. 
2. How about we say something like: `Support multiple salts for key rotation with AES-128-CBC. Only the first salt value is used for encryption. The additional salt values are used for decryption in sequential order if the previous decryption attempt fails.` 

    Note that I 
    * updated `key replacement` to `key rotation`.
    * differentiated salts and keys, as I believe they are not the same thing in this context.

Let me know what you think.

Reply:
1. I don't think we need to keep emphasising this algorithm `AES-128-CBC`, key replacement has nothing to do with the algorithm, only one key actually works, we want to replace the key when the key is no longer secure, the rest of the keys are used to ensure that the old encrypted data can still be decrypted properly. so i say `For the convenience of key replacement`
2. the key in this context is some kind of salt, you could search `iv` and `aes`, the key is the `iv`

Reply:
1. true but like you said "this encryption won't happen without a salt". There are algorithms that encrypt without a salt. Salts are optionally used to make things more secure. 
2. salt, key, and iv are three different things. see https://www.ibm.com/docs/en/imdm/12.0?topic=encryption-generating-aes-keys-password. The hex value here is the Salt. So why does it make sense to document salt with `For the convenience of key replacement, we support using multiple keys simultaneously, encrypting only using the first key, and using the keys in sequence to decrypt (if decryption fails).`

Reply:
It's not `salt`
```
aes:new(ivs, nil, aes.cipher(128, "cbc"), {iv = ivs}))
```
https://github.com/apache/apisix/blob/master/apisix/ssl.lua#LL81C29-L81C29

And the function prototypes
```
function _M.new(self, key, salt, _cipher, _hash, hash_rounds, iv_len, enable_padding)
```
https://github.com/openresty/lua-resty-string/blob/master/lib/resty/aes.lua#L120

As you see, it's used for the `key` and `iv`, not `salt`, the salt is `nil`

Reply:
## Discussion Summary

### if the salt is required

`the encryption won't happen if without a salt` is false. The encryption can take place without a salt. See the example in [lua-resty-string](https://github.com/openresty/lua-resty-string/tree/master):

```lua
local aes_256_cbc_with_padding = aes:new(key, nil, aes.cipher(256,"cbc"), {iv = string.sub(key, 1, 16)}, nil, nil, enable_padding)
-- the salt is the second argument which is nil
-- function _M.new(self, key, salt, _cipher, _hash, hash_rounds, iv_len, enable_padding)
```

The reason why it was communicated that way was because these hex values in the configs are passed as `iv` not `salt`. `iv` is required for the encryption to take place. 

### if the value is a key or a salt

It is indeed used as the `iv` in code, not the `salt`.

See how `keyring` is read from local_conf `apisix.data_encryption. keyring` and saved to `ivs`.
```lua
local function get_aes_128_cbc_with_iv_gde(local_conf)
    if _aes_128_cbc_with_iv_tbl_gde == nil then
        local ivs = core.table.try_read_attr(local_conf, "apisix", "data_encryption", "keyring")
        _aes_128_cbc_with_iv_tbl_gde = init_iv_tbl(ivs)
    end
    return _aes_128_cbc_with_iv_tbl_gde
end
```
and then ivs is passed into `aes:new(ivs, nil, aes.cipher(128, "cbc"), {iv = ivs})`:
```
local function init_iv_tbl(ivs)
    local _aes_128_cbc_with_iv_tbl = core.table.new(2, 0)
    local type_ivs = type(ivs)

    if type_ivs == "table" then
        for _, iv in ipairs(ivs) do
            local aes_with_iv = assert(aes:new(iv, nil, aes.cipher(128, "cbc"), {iv = iv}))
            core.table.insert(_aes_128_cbc_with_iv_tbl, aes_with_iv)
        end
    elseif type_ivs == "string" then
        local aes_with_iv = assert(aes:new(ivs, nil, aes.cipher(128, "cbc"), {iv = ivs}))
        core.table.insert(_aes_128_cbc_with_iv_tbl, aes_with_iv)
    end

    return _aes_128_cbc_with_iv_tbl
end
```

## Conclusion

I will update the comments for these values to "keys" to be consistent with the code as suggested by @monkeyDluffy6017. I don't however understand why the key for AES-128-CBC is not 128bit, mathematically speaking.